Add Module Setup and Teardown Callback to Linkable Code Interface (#145)
NVIDIA/numba-cuda: docs/source/user/cuda_ffi.rst

kernel as either PTX or LTOIR, depending on whether LTO is enabled. Other files
will be passed directly to the CUDA Linker.

For cumodules linked to a linkable code object, additional setup
and teardown callbacks are available. Setup functions are invoked once for

> every new module loaded. Teardown functions are invoked before the module is
> garbage collected by python. These callbacks are defined as follows:

Reviewer: I think the module is getting unloaded during the GC process or when the refcount has already dropped to zero outside of a reference cycle, so maybe the following is a closer representation of what happens:

```suggestion
every new module loaded. Teardown functions are invoked just prior to
module unloading. These callbacks are defined as follows:
```

.. code::

>   def setup_callback(

Author: Yes - I was also debating whether we should list two, though their signature is the same. Perhaps we should just do so.

>   def setup_callback(

Reviewer: I think that would be fine, yeah. Also, I realise I'm confused about how the "correct" stream is supposed to be selected or passed for the teardown callback, but maybe that will become clear with a bit more thinking / review.

>   def setup_callback(

Reviewer: Yes. Maybe teardown shouldn't take a stream? unloading a module must surely synchronize the device, so there seems not to be much point in trying to do the teardown on a stream. (I know we're not unloading the module now, but we *should* be, and ought to in future)

>   def setup_callback(

Reviewer: > IIRC we are unloading the module. It is `_module_finalizer` in driver.py no?

Ah, I see the disconnect here... I was thinking we don't unload it as soon as we can (which is when the kernel / jitted function that created it is no longer reachable) because the current context holds on to a reference to all modules. If the context were destroyed then we would probably unload the module, but I think that never happens in practice in any user code.

>   def setup_callback(

Reviewer: With your proposal nullified, does that mean you settled on the idea of not having teardown take a stream? Or just to think more about it for another way?

>   def setup_callback(

Author: The most recent version of the design dropped stream support for both setup and teardown function since this wasn't originally the ask of the feature request. And I don't want to over-engineer some unexpected stream behavior for callbacks.
